File: golang/1416.restore-the-array.go
const MOD int = 1e9 + 7


func numberOfArrays(s string, k int) int {
    n := len(s)
    dp := make([]int, n + 1)
    dp[n] = 1

    for i := n - 1; i >= 0; i-- {
        val := int(s[i] - '0')

        if val == 0 {
            continue
        }
        
        for j := i + 1; val <= k && j <= n; j++ {
            dp[i] = (dp[i] + dp[j]) % MOD
            if j < n {
                val = val * 10 + int(s[j] - '0')
            }
        }
    }
    
    return dp[0]
}

File: python/1416.restore-the-array.py
class Solution:
    def numberOfArrays(self, s: str, k: int) -> int:
        mod = 10**9 + 7
        n = len(s)
        dp = [0] * (n + 1)
        dp[n] = 1
        
        for i in reversed(range(n)):
            if s[i] == '0':
                continue

            j = i + 1
            count = 0
            while int(s[i:j], 10) <= k and j <= len(s):
                dp[i] += dp[j]
                j += 1
            dp[i] %= mod

        return dp[0]